Initial release of golang Skydive operator

oranichu opened the pull request:
This is a work in progress for the skydive-golang operator.
Right now the operator can create all the deployment services and routes needed for the analyzer and also the replica set for the agents.
Feel free to comment, correct and add whatever you think is releveant

Commit:
Skydiveanalyzer reconciler creats Deployment, service, and routes for both old and new UI.
Skydiveagnets reconciler creates DaemonSet (pods are stuck in pending)
Added README.md.

amorenoz commented:
cc @russellb

eranra commented on README.md:
```diff
@@ -0,0 +1,51 @@
+make manifests
+kubectl create -f config/crd/bases
+kubectl create -f config/samples/skydive_v1beta1_skydiveanalyzer.yaml 
+kubectl create -f config/samples/skydive_v1beta1_skydiveagents.yaml 
```
(1) change the name from skydive_v1beta1_skydiveanalyzer to skydive_v1beta1_skydive_analyzer and from skydive_v1beta1_skydiveagents to skydive_v1beta1_skydive_agents
(2) move outside of the samples folder

oranichu replied:
done

rcarrillocruz replied:
There is a mix of oc and kubectl commands.
My understanding is that we are trying to make this upstream first, thus I'd change the commands to just kubectl.

If there are specific options to make this to work on OpenShift, then it's worth having a README_OPENSHIFT.md

oranichu replied:
Pushed a script that deploys it on openshift
still working on making it kubectl friendly

eranra commented on README.md:
```diff
@@ -0,0 +1,51 @@
+kubectl create -f config/crd/bases
+kubectl create -f config/samples/skydive_v1beta1_skydiveanalyzer.yaml 
+kubectl create -f config/samples/skydive_v1beta1_skydiveagents.yaml 
+make run
```
Add infomration on printing more and less debug infomration

oranichu replied:
What did you mean here?

eranra replied:
@oranichu as we spoke on the phone ... manage the level of logs printed when executing `make run`

russellb commented:
Thanks for working on this!  One quick thing I noticed is that for this upstream operator, we can't do anything that is OpenShift specific, which includes the `Route` resource.  Maybe it'll be ok to include some optional OpenShift features that get used when the operator detects it's on OpenShift, but it'll need to work on non-OpenShift kubernetes clusters, as well.  I'd try it out on a `kind` cluster to validate that.

eranra commented on controllers/skydiveagents_controller.go:
```diff
@@ -0,0 +1,89 @@
+	assets := manifests.NewAssets("assets")
+
+	log.Info("Building configuration")
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfigPath)
```
Replace with

```	
var config *rest.Config
var err error
if kubeConfig := os.Getenv("KUBECONFIG"); kubeConfig != "" {
	config, err = clientcmd.BuildConfigFromFlags("", kubeConfig)
} else {
	config, err = rest.InClusterConfig()
}
```

oranichu replied:
So in my case, I'll have to add an ENV var named KUBECONFIG and set its path right?

eranra replied:
@oranichu yes or if you are logged into a cluster using kubectl in the current shell you should also be covered

Commit:
Combine skydive analyzer and skydive agents into skydive suite.
Add script for openshift easy deploy.
Fix kube config path error.
Delete redundant logs messages.

oranichu committed:
Delete api/v1beta1 directory

Old folder no longer needed

Commit:
Fix service issue on kins installation
Add instructions on how to run on kind

Commit:
Env variables are now passed through skydive suite yaml.
Easier access to modification of skydive analyzer, agents and flow exporter.

Commit:
Commit before skydivesuite name change to skydive

Commit:
change name to skydive from skydivesuite
add flow-exporter controller
add flow-exporter-dev to hack folder
scripts for run operator on kind, run flow-exporter and run flow-exporter-dev for development purposes using kind.

eranra commented on README.md:
```diff
@@ -0,0 +1,75 @@
+<!-- ABOUT THE PROJECT -->
+
+# Skydive Operator (WIP)
```
remove (WIP)

eranra commented on README.md:
```diff
@@ -0,0 +1,75 @@
+
+# Skydive Operator (WIP)
+
+This is a simple operator to deploy skydive analyzer and agents.
```
remove the word simple, add flow exporter

eranra commented on README.md:
```diff
@@ -0,0 +1,75 @@
+    * If you wish to run this locally by using kind please
+      check [KindInstallationREADME.md](hack/KindInstallationREADME.md)
+
+### Modifying the CRD
```
move after installation section (and you can call this section Customize skydive deployment / CRD)

eranra commented on README.md:
````diff
@@ -0,0 +1,75 @@
+run the script :
+
+```
+bash deploy_skydive_operator_on_openshift.sh
````
remvoe the word bash and add ./

eranra replied:
All add the text ... make sure that you are logged into open shift cluster

eranra commented on deploy_skydive_operator_on_openshift.sh:
```diff
@@ -0,0 +1,52 @@
+# delete skydive project if it exists (to get new frech deployment)
```
s/frech/fresh

eranra commented on deploy_skydive_operator_on_openshift.sh:
```diff
@@ -0,0 +1,52 @@
+oc_set_credentials() {
+  # analyzer and agent run as privileged container
+  oc adm policy add-scc-to-user privileged -z default
+  # analyzer need cluster-reader access get all informations from the cluster
```
s/informations/information

eranra replied:
s/get/to get

eranra commented on deploy_skydive_operator_on_openshift.sh:
```diff
@@ -0,0 +1,52 @@
+fi
+}
+
+# deoploy skydive
```
s/deoploy/deploy

eranra commented on deploy_skydive_operator_on_openshift.sh:
```diff
@@ -0,0 +1,52 @@
+
+oc_clear_skydive_crd() {
+SKYDIVE_SUITE_CRD=$(oc get skydive)
+if [ ! -z "$SKYDIVE_SUITE_CRD" ]; then
```
s/! -z/-n

eranra commented:
@oranichu I started the operator today and got::

```2021-03-14T16:12:40.456+0200    ERROR   setup   problem running manager {"error": "no matches for kind \"SkydiveFlowExporter\" in version \"skydive.example.com/v1\""}
```

You need to deploy the CRD for FlowExporter in the skydive deployment script (and vice versa)

After running 
`kubectl create -f config/crd/bases/skydive.example.com_skydiveflowexporters.yaml`

It works :-)

Commit:
returned to only one main
fix readme
fix error with creating crd

Commit:
Add Prometheus Connector kind and script to deploy it

Commit:
Fix problems in hack folder scripts

rcarrillocruz commented:
Apologies for the delay, I was meant to suggest how to put the OpenShift certs for OVN visualization:

https://github.com/openshift/cluster-network-operator/blob/master/bindata/network/ovn-kubernetes/ovnkube-node.yaml#L328

There you can see the configmap and secrets used, in case you want to put that on the openshift script.

eranra commented:
@oranichu  ... so .... you will need to download those / copy the infomration from the `openshift-ovn-kubernetes` namespace to the skydive deployment namespace (configmaps can't be shared between namespaces)  ... than use them in skydive operator here: https://github.com/skydive-project/skydive/blob/master/etc/skydive.yml.default#L178

@rcarrillocruz above correct ??? 

![image](https://user-images.githubusercontent.com/3080696/111910448-10ebfe00-8a6a-11eb-9bd7-a9a957df7f41.png)

rcarrillocruz commented:
That's right, since this is in out-of-band script what you could do is oc -n openshift-ovn-kubernetes get <cm|secret> <blah> -o yaml > ovn-<cm|secret>.yaml, then pipe it to create it on the skydive operator namespace, that could be a way of doing it.

eranra commented:
Need to add this to the script ::
-=--==-=-=-=-

`oc get configmap ovn-ca --namespace=openshift-ovn-kubernetes -o yaml | sed 's/namespace: openshift-ovn-kubernetes/namespace: skydive/' | oc  apply --namespace=skydive -f -`

`oc get secret ovn-cert --namespace=openshift-ovn-kubernetes -o yaml | sed 's/namespace: openshift-ovn-kubernetes/namespace: skydive/' | oc apply --namespace=skydive -f -`

`MASTER_IP=$(kubectl get node --selector='!node-role.kubernetes.io/master' -o jsonpath='{.items[0].status.addresses[?(@.type=="InternalIP")].address}')`

In my case `MASTER_IP` was 10.0.139.85

And I changed analyzer deployment yaml to include::

```
            - name: SKYDIVE_ANALYZER_TOPOLOGY_OVN_ADDRESS
              value: tcp:10.0.139.85:9641
            - name: SKYDIVE_ANALYZER_TOPOLOGY_OVN_CA
              value: /ovn-cert/tls.crt
            - name: SKYDIVE_ANALYZER_TOPOLOGY_OVN_KEY
              value: /ovn-cert/tls.key
            - name: SKYDIVE_ANALYZER_TOPOLOGY_OVN_CACERT
              value: /ovn-ca/ca-bundle.crt

          volumeMounts:
          - mountPath: /ovn-cert
            name: ovn-cert
          - mountPath: /ovn-ca
            name: ovn-ca

      volumes:
      - name: ovn-ca
        configMap:
          name: ovn-ca
      - name: ovn-cert
        secret:
          secretName: ovn-cert
          
 ```       

I still get this error in skydive analyzer log::


`skydive-analyzer-6f97576954-rwhs2: Failed to start: failed to connect: dial tcp 10.0.139.85:9641: connect: connection refused, retrying in background every 3s`


I changed manually the IP to `10.0.175.118` which is the POD IP (need to understand what IP we need to use, it looks like OVN is not exposing the NB so we can't hook to Node IP :-) 

Now the code connects but there is probably something wrong with OVN ... I see in skydive logs ::


```
2021-03-21T16:33:21.871Z	[34mINFO[0m	http/server.go:105 (*Server).Listen	skydive-analyzer-6c969c6d6f-bllpr: Listening on socket 0.0.0.0:8082
2021-03-21T16:33:21.885Z	[34mINFO[0m	client/election.go:104 (*MasterElector).start	skydive-analyzer-6c969c6d6f-bllpr: starting as the master for /elections/alert-server: skydive-analyzer-6c969c6d6f-bllpr
2021-03-21T16:33:21.897Z	[34mINFO[0m	k8s/cluster.go:54 (*clusterCache).addClusterNode	skydive-analyzer-6c969c6d6f-bllpr: Added cluster{Name: cluster}
2021-03-21 16:33:21.902222 I | ListDbs failureunexpected EOF
```



I also tried from with in the container to execute::

```curl -k --cert /ovn-cert/tls.crt --key /ovn-cert/tls.key --cacert /ovn-ca/ca-bundle.crt https://10.0.175.118:9641/```

but I get empty response from server :-(

eranra commented:
@rcarrillocruz can you look on the above and advice ... maybe I did something wrong in the configuration ... mixed something ??? of maybe there is something that needs more work in OVN code ???

@oranichu @russellb  ^^^ FYI

eranra commented:
@oranichu BTW this discussion with @rcarrillocruz and the work to enable OVN support should not block this PR from merge ... we should merge this PR and add OVN support as next phase of functionality against the operator.

@lebauce FYI, hope you agree.

rcarrillocruz commented:
We do expose the OVN NB with a service, but it's a headless service:

https://github.com/openshift/cluster-network-operator/blob/dfea1dc53ed82f5e4a4e85b7e8e863af3e4bd54a/bindata/network/ovn-kubernetes/005-service.yaml

This is how the CNO finds out the IP, by getting the svc endpoints:

https://github.com/openshift/cluster-network-operator/blob/dfea1dc53ed82f5e4a4e85b7e8e863af3e4bd54a/bindata/network/ovn-kubernetes/005-service.yaml

So the script should be doing something similar.

I'm swamped the next couple of weeks with end of release stuff, will try to find some time to debug it but cannot give you timelines.

eranra commented:
@rcarrillocruz thanks ... we will need to look on that later ... NP

Commit:
Added explanations on how to fix KUBECONFIG problem

jotak commented on hack/KindInstallationREADME.md:
```diff
@@ -0,0 +1,66 @@
+Install kind on your machine, use [this](https://kind.sigs.k8s.io/docs/user/quick-start/) quick start guide for help, 
+make sure that your have a running kind cluster and that kubectl commands refer to relevant context.
+
+After installing and running the Kind cluster, make sure the env var KUBECONFIG is set to the correct kubeConfig file,
+it usually defaults to $HOME/.kube/config, run this command:
```
I think this part is also relevant for the openshift install, not only Kind. I just run into the issue, with this error in logs : `unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined`. I had no KUBECONFIG set up (I'm connecting to a remote CRC with `oc`. It works fine but doesn't set up the relevant kubeconfig when I log in, so I had to pull it manually from my remote crc)

Commit:
the operator will now look for KUBECONFIG if not defined in $HOME/.kube/config
